### ISSUE TITLE
React router should be able to update just state using funcs, as alternative to mounting components

### ISSUE THREAD

**Issue by mweststrate** [ReactTraining/react-router#4726]
Hi @ryanflorence & @mjackson 

As promised, an issue to follow up on the [twitter discussion](https://twitter.com/mweststrate/status/831218911731445761) we were having a few weeks back.
Sorry for the delay! Couldn't find time earlier to write this down :)
Anyway, first of all; this is not an attack on react-router! It's just part of my everlasting quest to cleaner abstractions and basically food for though. I can imagine this does not fit in the philosophy of RR, so feel free to close it at any moment. It is basically an elaboration of statements made earlier.

Lately, at Mendix we have stopped using react-router because it didn't fit well in our philosophy on state management anymore. The essence of the issue is: react router updates the component tree, where, imho, it should update the state of the application.

To demonstrate that, in a recent workshop I have been using the following webshop application, that starts out as an almost static mock, and needs to be implemented and tested. The starting branch can be found [here](https://github.com/mweststrate/react-mobx-shop/tree/0_base), and the running example [here](http://book-shop-2017-02-17.surge.sh/).

In the initial wireframe of the app, the state of the application is modified [entirely programmatically](https://github.com/mweststrate/react-mobx-shop/blob/0_base/src/components/App.js#L44). There are no links, no routing, just plain onClicks and function invocations. Easy testable and, with flow / typescript, easy to strongly type. Navigating to a specific book is simply:

```
  openBookPage = (book) => {
    this.setState({
      page: "book",
      selectedBook: book
    })
  }
```

However, this model changes radially when we add react-router to the project. Because a few things need to change:

1. We define a route for `book/selected` book and a component responsible for rendering it
2. Instead of programmatically change the state of our application, we create links and need to encode our 'function invocations' as strings
2. The component responsible for rendering the currently open book now also needs to parse the very same string, make sure the proper book is fetched etc. 
3. Programmatically going to specific books is now harder; if we want to open a random book, we need to update the state and make sure a history item is pushed as well.

So imho two thing are suboptimal here. 
1. 'string level' programming
2. because react router mounts components, the React model of `view = f(state)` breaks; rather; because router mounts our components, the mount process of components starts to fill in our state. 

This in turn makes it hard to test the entire user flow of your app without the view layer (components). Which should be perfectly possible if the view truly was just a side effect free function of the state.

So instead of having the model:

1. URL is popped
2. Router mounts correct component
3. component interprets route params, updates state. (Downside: Component is very much aware of routing)
4. state update might trigger additional renders: (Downside: Hard to test state without firing component) hooks

We could transition to this model:

1. URL is popped
2. Router calls matching callback
3. Callback updates state
4. Components render the state
5. URL is updated according to the state

![img](https://cdn-images-1.medium.com/max/800/1*ksoMhK4GRob9QHbzzb0SBA.png)

I think the essence of why I am trying to express is best demonstrated here, with two diffs I used in the same workshop

1. [adding react-router](https://github.com/mweststrate/react-mobx-shop/compare/5_serversync...6_routing_react_router)
2. [adding manual routing](https://github.com/mweststrate/react-mobx-shop/compare/5_serversync...6_routing)

The latter diff is a bit bigger because of some utilities to parse urls etc (could have used yester instead), but the essence is that the connection between views is still programmatically organized, and the view state is entirely explicit, instead of depending on components being mounted that are caused by the router. 
The essence of this approach:

1. During app start a [history listener is created](https://github.com/mweststrate/react-mobx-shop/blob/c875b6b7d35ce9fef5015d34e8fe2d5ca865788b/src/index.js#L36) that simply calls a best-match function if a url is popped
2. [ViewStore](https://github.com/mweststrate/react-mobx-shop/blob/c875b6b7d35ce9fef5015d34e8fe2d5ca865788b/src/stores/ViewStore.js) captures all state that determines the current view (but could also be component state)
3. [App](https://github.com/mweststrate/react-mobx-shop/blob/c875b6b7d35ce9fef5015d34e8fe2d5ca865788b/src/components/App.js) just renders the viewstate, picks the correct components based on the current state
4. ViewStore also calculates what the [current url](https://github.com/mweststrate/react-mobx-shop/blob/c875b6b7d35ce9fef5015d34e8fe2d5ca865788b/src/stores/ViewStore.js#L16) should look like 
5. ...Which, upon change, is [pushed](https://github.com/mweststrate/react-mobx-shop/blob/c875b6b7d35ce9fef5015d34e8fe2d5ca865788b/src/index.js#L28) to the browser history

This results in a few interesting observations:

1. Routing is now only a concern at the 'edges' of our application. 
2. Components are completely routing unaware. No component hooks or parameter parsing. In fact, all components are function components now
3. Stores are routing / browser unaware as well
4. Routing itself is now largely view-framework agnostic; Besides a `Link` component there is nothing React specific
5. Routing is state management framework agnostic; this works well with MobX, but as far as I can see this would integrates with Redux as easily!

To demonstrate, here is a unit tests that covers an entire user flow, without need a browser (simulation) or mounting components. All logic can be tested headless, [tested](https://github.com/mweststrate/react-mobx-shop/commit/95a53d8b3743fb3b896ac88d3a09440c79cb9cee#diff-7737fb441d2c98cc8a7aa6b68513d593R7). The route parsing functions are not tested here, but those are trivial to test seperately as well.

I blogged earlier about this model [here](https://medium.com/@mweststrate/how-to-decouple-state-and-ui-a-k-a-you-dont-need-componentwillmount-cc90b787aa37) and actually several routers where implement according to this philosophy [mobx-react-router](https://www.npmjs.com/package/mobx-react-router), [yester](https://github.com/basarat/yester/) and some others. 

Thanks for reading. Curious about your opinions :)

**Comment by ryanflorence** [comment]
tl;dr but we've had this conversation 8000 times with redux people. Let's do a screenshare so I can show you how this model breaks the browser back/forward buttons. Unless you've figured something out that we haven't :) if I'm wrong we can reopen. I'll reach out in a Twitter DM to find a time to chat.

**Comment by neptunian** [comment]
@ryanflorence  if that is the case, how would one use react-modal with react-router?  i want to show the modal on a certain route, but the isOpen state is responsible for hiding and showing the modal.

**Comment by timdorr** [comment]
@neptunian This is possible. On v3, look at the Pinterest example. On v4, see here: https://reacttraining.com/react-router/web/example/modal-gallery

**Comment by neptunian** [comment]
@timdorr the v4 example is not using react-modal component, though.

**Comment by neptunian** [comment]
@timdorr perhaps i'll ask on the react-modal repo, thanks.

**Comment by timdorr** [comment]
Just use the same technique. This is the key bit:
```js
const isModal = !!(
      location.state &&
      location.state.modal &&
      this.previousLocation !== location // not initial render
    )
```

**Comment by neptunian** [comment]
@timdorr yes but the problem is you cannot mount the react-modal component based on a route, it mounts when the page is loaded.  otherwise the animation won't work because instead of react-modal showing and hiding its children it mounts and unmounts the whole component when you use routing.

**Comment by neptunian** [comment]
i was able to get it working by using the children function:
https://reacttraining.com/react-router/web/api/Route/children-func

because react-modal needs to mount whether a route matches or not and then show itself based on passing a state.
